torch/csrc/jit/passes/quantization.cpp

```diff
@@ -676,5 +677,255 @@ TORCH_API script::Module InsertObservers(
+      for (const Use& u: v->uses()) {
+        // Note that here we just check for the name of observer, but the ideally
+        // we should be comparing the type of observer, this is a temporary
+        // work around until data only clone of module.clone is supported.
+        auto submodule_name = u.user->inputs().at(0)->debugName();
+        if (u.user->kind() == prim::CallMethod && u.user->s(attr::name) == "forward" &&
+//            u.user->inputs().at(0)->type()->kind() == c10::ClassType &&
+            submodule_name.find("observer_for_") == 0) {
+          auto observer_module = module.find_module(submodule_name);
+          return std::make_tuple(submodule_name, observer_module);
+        }
+      }
+      return std::make_tuple("", c10::nullopt);
+}
+
+IValue getQParam(const script::Module& module, Value* v) {
+    TORCH_INTERNAL_ASSERT(v->type()->isSubtypeOf(TensorType::get()));
+    c10::optional<script::Module> observer_module;
+    std::tie(std::ignore, observer_module) = findObserverModule(module, v);
+    TORCH_INTERNAL_ASSERT(observer_module,
+                          "getQParam expects the corresponding observer for ",
+                          v->debugName(),
+                          " exists.");
+    auto calculate_qparams = observer_module.value().get_method("calculate_qparams");
+    IValue qparams = calculate_qparams(std::vector<IValue>());
+    return qparams;
+}
+
+void quantizeBias(const script::Module& module, Value* v) {
+  // Traverse to the place where this is used
+  Node* aten_call_node = nullptr;
+  std::vector<std::string> ops_with_bias = {"aten::conv2d", "aten::_convolution"};
+  for (const auto& use: v->uses()) {
+    if (std::find(ops_with_bias.begin(), ops_with_bias.end(),
+                  use.user->kind().toQualString()) != ops_with_bias.end()) {
+      Value* activation = use.user->inputs()[0];
+      Value* weight = use.user->inputs()[1];
+      // Get qparam from activation
+      IValue act_qparam = getQParam(module, activation);
+      // Get qparam from weight
+      IValue weight_qparam = getQParam(module, weight);
+      IValue bias_scale =  at::scalar_tensor(
+          c10::Scalar(act_qparam.toTuple()->elements()[0].toTensor().item().toDouble() *
+                      weight_qparam.toTuple()->elements()[0].toTensor().item().toDouble()),
+          at::kDouble);
+      IValue bias_qparam = c10::ivalue::Tuple::create(
+          std::vector<IValue>({bias_scale,
+                               at::scalar_tensor(c10::Scalar(0))}), act_qparam.toTuple()->type);
+      Node* dequant = insertQuantDeQuantCall(v, bias_qparam, at::kQInt32);
+      v->replaceAllUsesWith(dequant->output());
+      Node* q = traverseToQuantNode(dequant);
+      TORCH_INTERNAL_ASSERT(q != nullptr);
+      q->replaceInputWith(dequant->output(), v);
+      break;
+    }
+  }
+}
+
+void quantizeValue(const script::Module& module,
```

Comment:
can we rename. As we are both quantizing tensor and inserting q-dq node

Reply:
quantizing tensors does require us to calculate qparam for them, I know previously it's separate because we have a global observer module, but in the current setting when we have per tensor observer I think they can be just one step.

Reply:
ok sounds good